src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs

```diff
@@ -0,0 +1,234 @@
+                ifMatch,
+                GetCanonicalizedHeaders(request),
+                GetCanonicalizedResource(request.RequestUri, StorageAccount),
+                md5,
+                contentType
+                );
+
+            var SignatureBytes = Encoding.UTF8.GetBytes(MessageSignature);
+            var SHA256 = new HMACSHA256(Convert.FromBase64String(StorageKey));
+            var AuthorizationHeader = "SharedKey " + StorageAccount + ":" +
+                                      Convert.ToBase64String(SHA256.ComputeHash(SignatureBytes));
+            return AuthorizationHeader;
+        }
+
+        public static string GetCanonicalizedHeaders(HttpWebRequest request)
+        {
+            var headerNameList = new ArrayList();
+            var sb = new StringBuilder();
+            foreach (var headerName in request.Headers.AllKeys)
+            {
+                if (headerName.ToLowerInvariant().StartsWith("x-ms-", StringComparison.Ordinal))
+                {
+                    headerNameList.Add(headerName.ToLowerInvariant());
+                }
+            }
+            headerNameList.Sort();
+            foreach (string headerName in headerNameList)
+            {
+                var builder = new StringBuilder(headerName);
+                var separator = ":";
+                foreach (string headerValue in GetHeaderValues(request.Headers, headerName))
+                {
+                    var trimmedValue = headerValue.Replace("\r\n", string.Empty);
+                    builder.Append(separator);
+                    builder.Append(trimmedValue);
+                    separator = ",";
+                }
+                sb.Append(builder);
+                sb.Append("\n");
+            }
+            return sb.ToString();
+        }
+
+        public static ArrayList GetHeaderValues(WebHeaderCollection headers, string headerName)
+        {
+            var list = new ArrayList();
+            var values = headers[headerName];
+            if (values != null)
+            {
+                list.Add(values.TrimStart(null));
+            }
+            return list;
+        }
+
+        public static string GetCanonicalizedResource(Uri address, string accountName)
+        {
+            var str = new StringBuilder();
+            var builder = new StringBuilder("/");
+            builder.Append(accountName);
+            builder.Append(address.AbsolutePath);
+            str.Append(builder);
+            var values2 = new Dictionary<string, HashSet<string>>();
+            if (true)
+            {
+                var values = new Dictionary<string, HashSet<string>>();
+                var newreg = new Regex(@"\?(\w+)\=([\w|\=]+)|\&(\w+)\=([\w|\=]+)");
+                var matches = newreg.Matches(address.Query);
+                foreach (Match match in matches)
+                {
+                    string key, value;
+                    if (match.Groups[1].Value != String.Empty)
```

Comment:
I'd use "String.IsNullOrEmpty", the reference check to that const is probably going to be valid (I think we ALWAYS intern the empty string), but the other handles null, doesn't depend on that particular implementation detail, and is pretty standard "I don't want non-values" for strings.